File: test/csharp/SpecGateFixtures/AsyncFetch/async_fetch.cs
using SpecGate.Annotations;

namespace SpecGateFixtures.AsyncFetch;

/// <summary>
/// Async operation fixture — mirrors the Rust <c>fetch</c> async fn. The body
/// awaits an immediately-ready task so the <c>await</c> machinery is genuinely
/// exercised without requiring a reactor, proving the harness awaits async ops
/// and unwraps <see cref="Task{TResult}"/> for the result.
/// </summary>
public static class Fetcher
{
    /// <summary>Returns a canned response for <paramref name="url"/>.</summary>
    /// <param name="url">The URL to fetch (spec input <c>url</c>).</param>
    /// <returns>A response string for the given URL.</returns>
    [SpecOperation("fetch")]
    public static async Task<string> Fetch([SpecInput("url")] string url) =>
        await Task.FromResult($"response from {url}");
}

File: test/rust/crates/specgate-ctsc-fixtures/src/conformance/async/async_fetch.rs
use specgate::spec_operation;

#[spec_operation("fetch", spec = "fixture.async_fetch")]
pub async fn fetch(url: &str) -> String {
    std::future::ready(format!("response from {url}")).await
}

#[test]
fn async_operation_is_discovery_valid_without_native_capture() {
    let operation = specgate::__rt::SPECGATE_OPS
        .iter()
        .find(|operation| operation.component == "fixture.async_fetch" && operation.name == "fetch")
        .expect("async fetch metadata");
    assert!(operation.is_async);
    assert_eq!(operation.params, &[("url", "& str")]);
}
